presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java

```diff
@@ -0,0 +1,206 @@
+    {
+        this.channels = channels;
+        this.resultFuture = SettableFuture.create();
+        this.tupleDomainConsumer = this::addPartition;
+
+        this.partitions = new TupleDomain[partitionCount];
+        this.partitionCounter = 0;
+    }
+
+    private void addPartition(TupleDomain<DynamicFilters.Descriptor> tupleDomain)
+    {
+        TupleDomain<DynamicFilters.Descriptor> result = null;
+        synchronized (this) {
+            if (partitions == null) {
+                return;
+            }
+            if (tupleDomain.isAll()) {  // disable dynamic filtering - all build-side values are needed.
+                partitions = null;  // drop all collected predicates if one of partitions' results is too large.
+                result = TupleDomain.all();
+            }
+            else {
+                // Called by each DynamicFilterCollectOperator instance (when collection is over).
+                verify(partitionCounter < partitions.length);
+                partitions[partitionCounter++] = tupleDomain;
+                if (partitionCounter == partitions.length) {
+                    result = Arrays.stream(partitions).reduce(TupleDomain.none(), TupleDomain::columnWiseUnion);
+                    // NOTE: may result in a bit more relaxed constraint if there are multiple columns and multiple rows.
+                    // See the comment at TupleDomain::columnWiseUnion() for more details.
+                    partitions = null;  // drop collected predicates after collection is over.
+                }
+            }
+        }
+        if (result != null) {
+            verify(resultFuture.set(result), "dynamic filter result is provided more than once");
+        }
+    }
+
+    public static Optional<DynamicFilter> create(JoinNode planNode, TypeProvider types, int partitionCount)
+    {
+        if (!isInnerReplicatedJoin(planNode)) {
+            return Optional.empty();
+        }
+        Map<String, Symbol> dynamicFilters = planNode.getDynamicFilters();
+        PlanNode buildNode = planNode.getRight();
+        List<FilterNode> filterNodes = PlanNodeSearcher
+                .searchFrom(planNode.getLeft())
+                .where(DynamicFilter::isFilterAboveTableScan)
+                .recurseOnlyWhen(DynamicFilter::isInnerReplicatedJoin)
+                .<FilterNode>findAll();
+
+        ImmutableList.Builder<Channel> builder = ImmutableList.builder();
+        for (FilterNode filterNode : filterNodes) {
+            TableScanNode tableScanNode = (TableScanNode) filterNode.getSource();
+            DynamicFilters.ExtractResult extractResult = DynamicFilters.extractDynamicFilters(filterNode.getPredicate());
+            for (DynamicFilters.Descriptor descriptor : extractResult.getDynamicConjuncts()) {
+                Symbol probeSymbol = Symbol.from(descriptor.getInput());
+                if (!tableScanNode.getAssignments().containsKey(probeSymbol)) {
```

Comment:
that shouldn't happen. Otherwise dynamic filtering planning would be incorrect

Reply:
You're right, fixed.